File: OJ_Assignment/Ch06-Binary_Tree/1042.6.51.py
class TreeNode:
    def __init__(self, val = "", left = None, right = None):
        self.data  = val
        self.left  = left
        self.right = right

def OpHigher(op1, op2):
    return (op1 == '*' or op1 == '/') and (op2 == '+' or op2 == '-')

def SubLowOp(op1, op2):
    return (op1 == '-') and (op2 == '+' or op2 == '-')

i = 0
def CreateTree(tnode, str):
    global i
    c = str[i]
    i += 1
    if c == '#':
        return None
    else:
        tnode = TreeNode(c)
        tnode.left  = CreateTree(tnode.left, str)
        tnode.right = CreateTree(tnode.right, str)
        return tnode

def InOrder(tnode):
    if tnode is None:
        return
    flg = False

    if tnode.left is not None:
        flg = OpHigher(tnode.data, tnode.left.data)
    else:
        flg = False    
    if flg:
        print("(", end="")
        InOrder(tnode.left)
        print(")", end="")
    else:
        InOrder(tnode.left)
    print(tnode.data, end="")

    if tnode.right is not None:
        flg = OpHigher(tnode.data, tnode.right.data) or SubLowOp(tnode.data, tnode.right.data)
    else:
        flg = False
    if flg:
        print("(", end="")
        InOrder(tnode.right)
        print(")", end="")
    else:
        InOrder(tnode.right)

tree = None
line = input()
tree = CreateTree(tree, line)
InOrder(tree)
print()

File: OJ_Assignment/Ch06-Binary_Tree/1042.6.51.cpp
#include <iostream>
using namespace std;

struct tnode {
    char val;
    tnode *lft;
    tnode *rgt;
};

inline bool OpHigher(char op1, char op2) {
    return (op1 == '*' || op1 == '/') && (op2 == '+' || op2 == '-');
}

inline bool OpLowEqual(char op1, char op2) {
    return op1 == '-' && (op2 == '+' || op2 == '-');
}

void CreateTree(tnode *&ptr) {
    char c = cin.get();
    if (c == '#') {
        ptr = nullptr;
    } else {
        ptr = new tnode;
        ptr->val = c;
        CreateTree(ptr->lft);
        CreateTree(ptr->rgt);
    }
}

void InOrder(tnode *root) {
    if (root == nullptr) {
        return;
    }
    bool flg = false;

    if (root->lft != nullptr) {
        flg = OpHigher(root->val, root->lft->val);
    } else {
        flg = false;
    }
    if (flg) cout << "(";
    InOrder(root->lft);
    if (flg) cout << ")";

    cout << root->val;

    if (root->rgt != nullptr) {
        flg = OpHigher(root->val, root->rgt->val) || OpLowEqual(root->val, root->rgt->val);
    } else {
        flg = false;
    }
    if (flg) cout << "(";
    InOrder(root->rgt);
    if (flg) cout << ")";
}

int main() {
    tnode *ptree = nullptr;
    CreateTree(ptree);
    InOrder(ptree);
    cout << endl;
    return 0;
}
